### DISCUSSION TITLE
Custom usefetcher invalid hook error in nextjs typescript

### DISCUSSION THREAD

**Discussion by hbisadii** [vercel/next.js#44336] [Help]
[](url)I am trying to make a custom usefetcher in my Nextjs18 typescript project but it keeps giving me the "Invalid hook call. Hooks can only be called inside of the body of a function component." error. Here is the code:
```
import { Axios as request } from "./axios";
import { AxiosRequestConfig } from "axios";
import { useContext } from "react";
import { UserContext } from "../../context";

interface IConfig extends AxiosRequestConfig {
  useToken?: boolean;
  formData?: boolean;
}

export const UseFetcher = <T, G = any>(config: IConfig) => {

  const axiosConfig: AxiosRequestConfig = config;
  //token and user data
  const {state: userState} = useContext(UserContext);
  //header config
   axiosConfig.headers = {
    ...axiosConfig.headers,
    accept: "application/json",
    "x-http-request-details": "app-version-code=1;platform=0;",
    appversion: "1",
    "Content-Type": config?.formData
      ? "multipart/form-data"
      : "application/json;charset=UTF-8",
    ...(config.useToken
      ? { Authorization: `Bearer ${userState?.tokens?.access}` }
      : null),
  };

  return (data?: G) =>
    new Promise<T>((resolve, reject) => {

      if (data) axiosConfig.data = data;
      if (data && axiosConfig.method === "GET") axiosConfig.params = data;
      //send request
      return request(axiosConfig)
        .then((res) => {
          //handle success
          if (res?.data?.isSuccess) resolve(res?.data);
          else {
            //handleFailedRequest(res as any);
            //reject(res?.data);
            // handleFailedRequest(res)
          }
        })
        .catch((error) => {
          //handle unsuccess
          //handleFailedRequest(error);
          //reject(error?.response?.data);
        });
    });
};

```
![Screen Shot 2022-12-25 at 12 43 15 AM](https://user-images.githubusercontent.com/120975133/209458123-bd772932-b6f4-4054-8c6b-6a47c0add72f.png)
I have "@types/react": "^18.0.26", "@types/react-dom": "^18.0.9", "react": "18.2.0", "react-dom": "18.2.0", but I also tried deleting node_modules and yarn.lock and running npm script for installing the latest versions which didn't work.

**Comment by icyJoseph** [comment]
How are you calling this hook? If possible share a repository please. The last bit of your screenshot seems to suggest that, the hook is invoked within a file called `CheckNumber.tsx`. Note that custom hooks ought to be called within the React tree. 

Is `src/services/authentication` something that happens within the React tree?

> Also, I'd recommend, renaming `UseFetcher` to `useFetcher`.
